### PR TITLE
1916 change log crate 

### LINKED ISSUE #1916: Change log crate
## Is your feature request related to a problem? Please describe 👀
Missing initial logs since logging function was moved to allow log level to be updated since fast_log can only be initialised once. Also see https://github.com/openmsupply/open-msupply/pull/1897#discussion_r1228834457

## Describe the solution you'd like 🎁
Use a log that can be updated during runtime, maybe https://docs.rs/simple-log/latest/simple_log/ ?

### PR DESCRIPTION
**Pull request by roxy-dao**
Fixes #1916

# 👩🏻‍💻 What does this PR do? 
Change the log crate as suggested [here](https://github.com/openmsupply/open-msupply/pull/1897#discussion_r1228834457). This crate allows `log_level` to be updated at runtime, though server will still need to be restarted. Though, the LogSettings can be saved to `key_value_store` and `log_level` update can be moved out of the `start_server` method to allow `log_level` to update without having to restart the server. 

When the log file has exceeded the file size, it is compressed, so have added a check in `get_log_content` to decompress the file to allow content to be fetched.

# 🧪 How has/should this change been tested? 
- Probably best to use a bigger data file so logs get split, then call log endpoints to see information such as log_files/contents.
- Test on Android as well. Initialise on Android -> Go to Admin -> Show log. Log should still be there.

### CODE REVIEW

File: server/android/src/android.rs [hunk 1]
```diff
@@ -24,23 +22,6 @@ pub mod android {
 
     static SERVER_BUCKET: Mutex<Option<ServerBucket>> = Mutex::new(None);
 
-    struct AndroidLogger {}
```
Comment (roxy-dao) [inline server/android/src/android.rs:27]: Not sure the reasoning behind logging twice in standard oms log file and in android logcat... from what I can see, simple_log doesn't have the functionality to log to android logcat 🤔

File: server/service/Cargo.toml [hunk 2]
```diff
@@ -29,13 +29,14 @@ tera = "1"
 tokio = { version = "1.17.0", features = ["macros", "sync", "time"] }
 headless_chrome = "1.0.5"
 pretty_assertions = "1.3.0"
+flate2 = "1.0.26"
```
Comment (roxy-dao) [inline server/service/Cargo.toml:32]: Needed crate to decompress zip

File: server/windows/src/windows.rs [hunk 3]
```diff
@@ -65,7 +65,7 @@ mod omsupply_service {
                 return;
             }
         };
-        logging_init(settings.logging.clone(), None);
+        logging_init(settings.logging.clone(), None, true);
```
Comment (andreievg) [inline server/windows/src/windows.rs:68]: Was wondering, why this one is 'update'

File: server/server/src/lib.rs [hunk 4]
```diff
@@ -119,7 +119,7 @@ pub async fn start_server(
             .unwrap();
     }
 
-    logging_init(settings.logging.clone(), None, log_level);
+    logging_init(settings.logging.clone(), log_level, true);
```
Comment (andreievg) [inline server/server/src/lib.rs:122]: Any reason why this one is config update ?

File: server/android/src/android.rs (hunk 1 above)
Comment (andreievg) [reply server/android/src/android.rs:27]: Hmm, can you please specify how we can look at these logs in emulator or on device when running app in debug mode ?

I found logcat logs super useful, and I think we should describe an alternative to inspect logs, when console logs are not available, i.e. in emulator (and possible on physical device), you can go to browse file through Android Studio, device manager, then `/data/user/0/org.openmsupply.client/files/remote_server.log`

File: server/service/src/log_service.rs [hunk 5]
```diff
@@ -31,9 +36,17 @@ pub trait LogServiceTrait: Send + Sync {
             Some(file_name) => file_name,
             None => default_filename,
         };
-
         let log_file_path = log_dir_path.join(&file_name);
-        let log_file_content = fs::read_to_string(log_file_path)?;
+
+        let log_file_content = if file_name.ends_with(".gz") {
```
Comment (andreievg) [inline server/service/src/log_service.rs:41]: Hmm I wonder if we need to test this on Android ? 

I did a quick check and simple-log uses log4rs, which uses flate2 or compression, I see you also use flate2, it looks like it should be fine for Android (it's just rust code)

File: server/android/src/android.rs (hunk 1 above)
Comment (andreievg) [reply server/android/src/android.rs:27]: Do we need android-log crate still ?

File: server/server/src/logging.rs [hunk 6]
```diff
@@ -1,39 +1,35 @@
...
-use log::LevelFilter;
 use service::settings::{Level, LogMode, LoggingSettings};
+use simple_log::LogConfigBuilder;
 
-pub fn logging_init(
-    settings: Option<LoggingSettings>,
-    apply_config: Option<Box<dyn Fn(LogConfig) -> LogConfig>>,
-    level: Option<Level>,
-) {
+pub fn logging_init(settings: Option<LoggingSettings>, level: Option<Level>, update: bool) {
```
Comment (andreievg) [inline server/server/src/logging.rs:6]: Would just add a quick comment somewhere in this method that simple-log is based on log4rs, which we can use if we need to extend logging functionality beyond what's available for simple-log

**Review by andreievg — CHANGES_REQUESTED** [review]
Look pretty good, a couple of request as per comments below:

* Remove `All` type
* Document how to look at logs on Android

I was wondering do you need to use this now somewhere in API (where logging level is set at runtime) ? Would assume this crate was added to do this configuration at runtime ?

File: server/windows/src/windows.rs (hunk 3 above)
Comment (roxy-dao) [reply server/windows/src/windows.rs]: Nice catch XD [cbd8bf8](https://github.com/openmsupply/open-msupply/pull/1921/commits/cbd8bf8502926e3be1be0cf5161f62a0590869a7)

File: server/server/src/logging.rs [hunk 7]
```diff
@@ -1,39 +1,35 @@
...
+    let config = match settings.mode {
+        LogMode::File => file_logger(&settings)
+            .level(log_level.to_string())
+            .output_file()
+            .build(),
+        LogMode::Console => LogConfigBuilder::builder()
+            .level(log_level.to_string())
+            .output_console()
+            .build(),
+        LogMode::All => file_logger(&settings).level(log_level.to_string()).build(),
```
[the remark this answers is not in the document]
Comment (roxy-dao) [reply server/server/src/logging.rs:22]: It can out in both file and console (:

File: server/android/src/android.rs (hunk 1 above)
Comment (roxy-dao) [reply server/android/src/android.rs:27]: Yes, will need android-log crate still if we want logcat 🤔

File: server/server/src/lib.rs (hunk 4 above)
Comment (roxy-dao) [reply server/server/src/lib.rs:122]: An endpoint to update log level was added [here](https://github.com/openmsupply/open-msupply/issues/1890) and the UI changes will be implemented [here](https://github.com/openmsupply/open-msupply/issues/1891). Will add comment...

Though I was thinking... We could save the log settings into `key_value_store` and have the log level updated when updateLogLevel is called... that way we don't need to tell the user to restart the server for log changes... What do you think? 🤔

File: server/server/src/logging.rs (hunk 6 above)
Comment (roxy-dao) [reply server/server/src/logging.rs:8]: [cbd8bf8](https://github.com/openmsupply/open-msupply/pull/1921/commits/cbd8bf8502926e3be1be0cf5161f62a0590869a7)

File: server/android/README.md [hunk 8]
```diff
@@ -1,25 +1,24 @@
...
 | aarch64-linux-android   | arm64-v8a   |
 | armv7-linux-androideabi | armeabi-v7a |
 
-* Add targets to to rust `rustup target add aarch64-linux-android && rustup target add armv7-linux-androideabi`
-* Tell rust to use pre built linker from NDK. This is defined in .cargo/config.toml, underd [target.*]. We need to add prebuild Android ABI from NDK to PATH so that linkers/ar files can be found. 
-   * Find your NDK location, my one is `~/Library/Android/sdk/ndk/22.1.7171670/toolchains/llvm/prebuilt/darwin-x86_64/bin/`
-   * Permanently add NDK_BIN env variable `echo "export NDK_BIN=~/Library/Android/sdk/ndk/22.1.7171670/toolchains/llvm/prebuilt/darwin-x86_64/bin/" >> ~/.zshrc`
-   * Open new terminal and validate $NDK_BIN is set correctly: `ls $NDK_BIN` (should give a list of files starting with aarch64 and armv7a)
-   * Now we can add `$NDK_BIN` to path when buiding: `PATH=$PATH:$NDK_BIN cargo build --release`
+- Add targets to to rust `rustup target add aarch64-linux-android && rustup target add armv7-linux-androideabi`
```
Comment (roxy-dao) [inline server/android/README.md:16]: ... happened because of prettier can change back if old version is preferred

**Review by andreievg — APPROVED** [review]
Thanks Roxy !

I've noticed a couple of other things, but I think we can do this in another issue:

* We only really need to store settings to database that can be updated via UI (logging settings can technically be saved in service provider and then provided through service context)
* Afaik, off is not one of the levels in simple_log
* Can use serde for Level (from/to string when saving in DB)
* logging_init to happen in service_start (I think we can consolidate this, between android/windows and standard version)

**Pull request merged**